plugins/woocommerce/src/Blocks/Domain/Services/CheckoutLink.php

```diff
@@ -0,0 +1,128 @@
+	 */
+	public function add_checkout_link_endpoint() {
+		add_rewrite_rule( '^checkout-link$', 'index.php?checkout-link=true', 'top' );
+	}
+
+	/**
+	 * Add the checkout link query var.
+	 *
+	 * @param array $vars The query vars.
+	 * @return array The query vars.
+	 */
+	public function add_query_vars( $vars ) {
+		$vars[] = 'checkout-link';
+		return $vars;
+	}
+
+	/**
+	 * Handle the checkout link endpoint.
+	 *
+	 * @return void
+	 */
+	public function handle_checkout_link_endpoint() {
+		if ( ! get_query_var( 'checkout-link' ) ) {
+			return;
+		}
+		wp_safe_redirect( $this->get_checkout_link() );
+		exit;
+	}
+
+	/**
+	 * Process the query params and return the checkout link to redirect to complete with session token.
+	 *
+	 * @return string The checkout link.
+	 */
+	protected function get_checkout_link() {
+		$controller = new CartController();
+		$controller->empty_cart();
+
+		// Populate cart with products.
+		$products = array_filter( explode( ',', wc_clean( wp_unslash( $_GET['products'] ?? '' ) ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		foreach ( $products as $product_id_qty ) {
+			if ( strpos( $product_id_qty, ':' ) !== false ) {
+				list( $product_id, $qty ) = explode( ':', $product_id_qty );
+			} else {
+				$product_id = $product_id_qty;
+				$qty        = 1;
+			}
+			if ( ! absint( $product_id ) ) {
+				continue;
+			}
+			try {
+				$controller->add_to_cart(
+					[
+						'id'       => absint( $product_id ),
+						'quantity' => absint( $qty ),
+					]
+				);
+			} catch ( \Exception $e ) {
+				wc_add_notice( $e->getMessage(), 'error' );
+			}
+		}
+
+		// Apply coupon if provided.
+		$coupon = wc_format_coupon_code( wp_unslash( $_GET['coupon'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+
+		if ( wc_coupons_enabled() && ! empty( $coupon ) ) {
+			try {
+				$controller->apply_coupon( $coupon );
+			} catch ( \Exception $e ) {
+				wc_add_notice( $e->getMessage(), 'error' );
+			}
+		}
+
+		// Nothing was added to the cart. We need to redirect to the cart page with an error notice. Since guests may not
+		// have a session, add the notice in the query string.
+		if ( wc()->cart->is_empty() ) {
+			$empty_cart_notice = __( 'The provided checkout link was out of date or invalid. No products were added to the cart.', 'woocommerce' );
+			wc_add_notice( $empty_cart_notice, 'error' );
+
+			if ( ! wc()->session->has_session() ) {
+				return add_query_arg( 'wc_error', $empty_cart_notice, wc_get_cart_url() );
+			}
+
+			return wc_get_cart_url();
+		}
+
+		$redirect_url = wc_get_checkout_url();
+
+		// If the user is logged in, the session is tied to the user ID. Do not use a cart token.
+		if ( ! is_user_logged_in() ) {
+			$session_token = CartTokenUtils::get_cart_token( wc()->session->get_customer_id() );
+			$redirect_url  = add_query_arg( 'session', $session_token, $redirect_url );
```

Comment:
One thing we discussed in slack is that instead of creating a new link, we should preserve query params, can we copy all query params to the new link? I think that's doable, if not, replace the `checkout-link` with `checkout` (or whatever is called), remove `products` and `coupon` and add `session`.

Reply:
https://github.com/woocommerce/woocommerce/pull/58140/commits/abd0e06da54cbd16e9cf4390cef8ccac911da4f2